src/calliope/model.py

```diff
@@ -567,3 +574,181 @@ def _recalculate_storage_initial(self, results: xr.Dataset) -> xr.DataArray:
+        results.attrs["termination_condition"] = ",".join(
+            set(result.attrs["termination_condition"] for result in results_list)
+        )
+
+        return results
+
+    def _spores_update_model(
+        self,
+        baseline_results: xr.Dataset,
+        all_previous_results: list[xr.Dataset],
+        spores_config: config_schema.SolveSpores,
+    ):
+        """Assign SPORES scores for the next iteration of the model run.
+
+        Algorithms applied are based on those introduced in <https://doi.org/10.1016/j.apenergy.2023.121002>.
+
+        Args:
+            baseline_results (xr.Dataset): The initial results (before applying SPORES scoring)
+            all_previous_results (list[xr.Dataset]):
+                A list of all previous iterations.
+                 This includes the baseline results, which will be the first item in the list.
+            spores_config (config_schema.SolveSpores):
+                The SPORES configuration.
+        """
+
+        def _score_integer() -> xr.DataArray:
+            """Integer scoring algorithm."""
+            previous_cap = latest_results["flow_cap"].where(spores_techs)
+
+            # Make sure that penalties are applied only to non-negligible deployments of capacity
+            min_relevant_size = spores_config.score_threshold_factor * previous_cap.max(
+                ["nodes", "techs"]
+            )
+
+            new_score = (
+                # Where capacity was deployed more than the minimal relevant size, assign an integer penalty (score)
+                previous_cap.where(previous_cap > min_relevant_size)
+                .clip(min=1, max=1)
+                .fillna(0)
+                .where(spores_techs)
+            )
+            return new_score
+
+        def _score_relative_deployment() -> xr.DataArray:
+            """Relative deployment scoring algorithm."""
+            previous_cap = latest_results["flow_cap"].where(spores_techs)
+            relative_cap = previous_cap / self.inputs["flow_cap_max"].where(
+                spores_techs
+            )
+
+            new_score = (
+                # Make sure that penalties are applied only to non-negligible relative capacities
+                relative_cap.where(relative_cap > spores_config.score_threshold_factor)
+                .fillna(0)
+                .where(spores_techs)
+            )
+            return new_score
+
+        def _score_random() -> xr.DataArray:
+            """Random scoring algorithm."""
+            previous_cap = latest_results["flow_cap"].where(spores_techs)
+            new_score = (
+                previous_cap.fillna(0)
+                .where(previous_cap.isnull(), other=np.random.rand(*previous_cap.shape))
+                .where(spores_techs)
+            )
+
+            return new_score
+
+        def _score_evolving_average() -> xr.DataArray:
+            """Evolving average scoring algorithm."""
+            previous_cap = latest_results["flow_cap"]
```

Comment:
In all other methods, it is always: `previous_cap = latest_results["flow_cap"].where(spores_techs)`. In 'evolving_average', the `.where(spores_techs)` bit is gone. Is that on purpose?

Reply:
It's just applied later on

Reply:
Ok, that's what I figured. I guess it saves some lines of code to do it only on the final product in this case. But could this also be potentially less efficient if you must do the averaging, subtraction and division operations for larger xarray objects? Imagine a case in which you have a humongous model, but only do spores for a couple of techs.

Reply:
These are quite efficient operations as they're just standard arithmetic applied to known types, so shouldn't be an issue. We can revisit if it does seems slow.